src/DFAMin.hs

```diff
@@ -35,6 +34,26 @@ import qualified Data.List as List
 --      end;
 -- end;
 
+-- observation : Q is always subset of P
+-- let R = P \ Q. then following algorithm is the equivalent of the Hopcroft's Algorithm
+--
+-- R := {{all nonaccepting states}};
+-- Q := {{all accepting states}};
+-- while (Q is not empty) do
+--      choose a set A from Q
+--      remove A from Q and add it to R
+--      for each c in ∑ do
+--           let X be the set of states for which a transition on c leads to a state in A
+--           for each set Y in R for which X ∩ Y is nonempty and Y \ X is nonempty do
+--                replace Y in R by the greater of the two sets X ∩ Y and X \ Y
+--                add the smaller of the two sets to Q
+--           end;
+--           for each set Y in Q for which X ∩ Y is nonempty and Y \ X is nonempty do
+--                replace Y in Q by the two sets X ∩ Y and X \ Y
+--           end;
```

Comment:
I get that when the smaller one goes in Q, both go in P, now only the bigger one goes in R. But in the second loop now we are mutating Y a second time? This relationship of this is less clear.

OK now that i right this is is starting to become clear. Maybe make note that the newly-inserted smaller set into Q does not contradict `Q` by construction? So there is some omnipotence going on that makes the second part the same whether we're iterating through `Q_before` or `Q_current`?

Reply:
If a set `Y1` was inserted to `Q` at first loop, then either `Y1 ∩ X` or `Y1 \ X` must be an empty set since `Y1` is `Y ∩ X` or `Y \ X`, so the guard condition of second loop must fail.
Therefore, we don't need to think about `Q_before` or `Q_after`.

Reply:
Yeah exactly. So if you could just put that somewhere in comments for the next person, would be much appreciated.